File: consistency-store/src/main/java/de/tudarmstadt/consistency/store/Store.java
package de.tudarmstadt.consistency.store;

/**
 * Models any kind of store. Values that are indexed by a key. They can be accessed by creating handles
 * that make it posssible to access a single database object. Handles are retrieved by specifying that key and a
 * consistency level.
 *
 *
 * @author Mirko Köhler
 */
public interface Store<Key, Service extends ReferenceService<Key>> {

	void commit(Transaction<Service> actions, Class<?> isolationLevel) throws Exception;


}

File: consistency-store/src/main/scala/de/tudarmstadt/consistency/store/scala/Store.scala
package de.tudarmstadt.consistency.store.scala

import scala.language.higherKinds

/**
	* Created on 08.08.18.
	*
	* @author Mirko Köhler
	*/
trait Store[Key, Val, Op[_,_] <: Operation[_,_]] {

	type Context <: SessionContext[Key, Val, Op]

	type Session[R] = Context => R

	def commit[R](session: Session[R]) : R =
		session(newSessionContext())

	def newSessionContext() : Context

	def close(): Unit = {
		//Do nothing by default
	}

}
